src/odemis/driver/smaract.py

```diff
@@ -3084,3 +3085,1139 @@ def SA_CTL_Move(self, handle, ch, pos_pm, _):
+        # If during startup a call to .reference is made, the referencing future will be added to the
+        # executor and be carried out right after the initialization referencing is done.
+        # In conclusion, no special handling is required during startup, although it does not make much sense
+        # to use the driver until the startup is complete. The position getter will log a warning in this case.
+        self.state._set_value(model.ST_STARTING, force_write=True)
+
+        # Referencing
+        # There are two options for referencing: the full referencing procedure and the shorter validation
+        # procedure. The validation procedure reuses the referencing parameters stored in memory. Afterwards,
+        # the system is ready to use and might produce reasonable position values. However,
+        # it is recommended to reference (i.e. adjust) the system frequently (at least during startup)
+        # since even small changes in the experimental setup or environment might cause the old
+        # adjustment parameters to become non-ideal. Therefore, the values for all channels in .referenced
+        # will be False, until the full referencing routine has been performed.
+        # These procedure can take a while (up to 10 minutes), especially after the power of the system
+        # has just been turned on. Therefore, we don't wait until the procedure is complete.
+        # Starting from standby is generally much faster.
+        self._executor = CancellableThreadPoolExecutor(1)  # one task at a time
+        channel_ref = {ch: False for ch in self.channels}
+        self.referenced = model.VigilantAttribute(channel_ref, readonly=True)  # VA dict str(channel) -> bool
+        if ref_on_init:
+            f = self.reference()
+        else:
+            f = self._validate()
+            logging.warning("Picoscale is not referenced. The reported positions might not be accurate "
+                            "until referencing occurs.")
+        f.add_done_callback(self._on_referenced)
+
+    def terminate(self):
+        # should be safe to close the device multiple times if terminate is called more than once.
+        self._polling_thread.cancel()
+        if self._executor:
+            self.stop()
+            self.core.SA_SI_Close(self._id)
+            self._executor = None
+        super(Picoscale, self).terminate()
+
+    @staticmethod
+    def scan():
+        """
+        Find all available Picoscale controllers.
+        returns: (set of str) locator strings
+        """
+        core = SA_SIDLL()
+        b_len = 1024
+        buf = create_string_buffer(b_len)
+        core.SA_SI_EnumerateDevices(c_char_p(b""), buf, byref(c_size_t(b_len)))
+        locators = buf.value.decode('latin1').split("\n")
+        return set(locators)  # sometimes the same locator is returned multiple times, convert to set for unique values
+
+    @staticmethod
```

Comment:
Fine. Note that an alternative would be to have a subclass of Picoscale (eg _PicoscaleScanned) with a much simpler __init__(), so that you directly read the name and channel numbers using the standard wrapper functions.